Group Console and AnsiColor entities and add usage examples

janekdb opened the pull request:
- Group AnsiColor and Console members
- Add Console IO redefinition examples
- Add Console control code examples
- Add color swatches to AnsiColor
- Add AnsiColor control code examples
- Link to StdIn for reading
- Fix link syntax for java.util.Formatter
- Fix withErr method example
### Reviewing

The matching Scaladoc is available here to make reviewing this easier,
- http://janekdb.github.io/scala/PR-5173/scala/Console$.html
- http://janekdb.github.io/scala/PR-5173/scala/io/AnsiColor.html

Happy to take reviews from anyone with the inclination!
### Comment Format

`AnsiColor` follows the commenting style here: http://docs.scala-lang.org/style/scaladoc.html whereas `Console` sticks to the older style because this had a greater presence in `Console`.
### IO Redefinition versus IO Redirection

I wasn't convinced the mechanism for substituting alternative streams in `Console` qualified as redirection but I'm open to persuasion.

janekdb commented:
Low bit depth previews. Please use links above for full versions.

![scaladoc-console-pr-5173-v1](https://cloud.githubusercontent.com/assets/1123855/15338176/a19a1c82-1c76-11e6-9715-705a573c40cc.png)
![scaladoc-ansicolor-pr-5173-v1](https://cloud.githubusercontent.com/assets/1123855/15338184/a7ba5582-1c76-11e6-9013-6cf873b5fe7e.png)

janekdb commented:
![scaladoc-console-pr-5173-v3](https://cloud.githubusercontent.com/assets/1123855/15338351/7044c668-1c77-11e6-9d46-20777834eacf.png)


![scaladoc-ansicolor-pr-5173-v2](https://cloud.githubusercontent.com/assets/1123855/15338306/475e6632-1c77-11e6-86de-eb2d86561f34.png)

janekdb commented on src/library/scala/io/AnsiColor.scala:
```diff
@@ -1,52 +1,164 @@
+    **/
   final val BLACK      = "\u001b[30m"
-  /** Foreground color for ANSI red */
+  /** Foreground color for ANSI red
```
Alignment. This needs fixing here and in all other cases.

janekdb commented:
/rebuild

janekdb commented:
Error was `*** Error in `/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java': free(): invalid pointer: 0x00000001003939a8 ***`

som-snytt commented on src/library/scala/io/AnsiColor.scala:
```diff
@@ -1,52 +1,163 @@
+  *
+  * ==ANSI Style and Control Codes==
+  *
+  * This group of escape codes provides control over text styling. For example turn on reverse video with bold and
```
"For example, to turn..."

som-snytt commented on src/library/scala/io/AnsiColor.scala:
```diff
@@ -1,52 +1,163 @@
+  *
+  * object ColorDemo extends App {
+  *
+  *   println(s"${REVERSED}${BOLD}Hello 1979!${RESET}")
```
I would have said "Party like it's 1979" but that's just me.

som-snytt commented:
Is there any sentiment for aliases `Red` for `RED` and `Blink` for `BLINK`?  I renamed on import the couple of times I used it. I meant to contribute aliases with deprecations a long time ago. All caps really does feel like 1979.

soc commented on src/library/scala/Console.scala:
```diff
@@ -12,12 +12,115 @@ import java.io.{ BufferedReader, InputStream, InputStreamReader, OutputStream, P
+ *        if (prime)
+ *          Console.println(s"${RESET}${GREEN}yes${RESET}")
+ *        else
+ *          Console.err.println(s"${RESET}${YELLOW_B}${RED}${UNDERLINED}NO!${RESET}")
```
Can't we drop the curly braces?

janekdb replied:
Sure. I'll remove as many as possible.

janekdb replied:
@soc - Now with minimal curly braces.

![scaladoc-pr-5173-without-curlies](https://cloud.githubusercontent.com/assets/1123855/15408912/524cfb78-1e0a-11e6-9fb4-8622f9dbe76a.png)

Initial version with Curlies<sup>MAX</sup>

![scaladoc-pr-5173-with-curlies](https://cloud.githubusercontent.com/assets/1123855/15409110/50313bc8-1e0b-11e6-91e0-3dc14edf2671.png)

I prefer the initial version because it's easier to pick out the codes. WDYT?

janekdb commented:
Changes made and generated docs push to GitHub pages. Preview not updated.

janekdb commented:
@som-snytt I'd got no particular appetite for `Red` and `Blink` but very happy to see these extensions added if you want to. We can drop `RED` etc out of the groups and mark as deprecated.

som-snytt commented:
Since naming is hard, I don't have a good way to improve `RED_B`. Cleanest might be a separate name space, such as
```
object background { final val Red = RED_B }
```
or
```
object Red { final val background = RED_B ; override final val toString = RED }
```
or a `background` macro extension method that automates `Red.background` for the color constants.

It's harder to pick up the constants from `object Red`, where you'd want your `s"${Red}Hi, $name${Reset}"` or similar to be constant folded, whether by `Red + "Hi"` or constant-aware `f"${Red}Hi"`.

Although I don't like that `RED_B` is all caps and has a suffix that isn't instantly mnemonic, I don't want to create needless work, especially since the new color bars are so pretty.

som-snytt commented on src/library/scala/Console.scala:
```diff
@@ -12,12 +12,115 @@ import java.io.{ BufferedReader, InputStream, InputStreamReader, OutputStream, P
+ *        val prime = (2 to candidate - 1).forall(candidate % _ != 0)
+ *
+ *        if (prime)
+ *          Console.println(s"$RESET${GREEN}yes$RESET")
```
I happen to find the uniform braces easier to read. Another option:

```
def greenly(s: String) = s"$RESET$GREEN$s$RESET"
println(greenly("yes"))
```

janekdb replied:
@soc - Having tried out brace minimisation I'm keen to restore the fully braced strings. I feel this makes the embedded constants stand out more which is useful in this context. What do you say?

lrytz replied:
+1 for fully curly braced strings here

lrytz commented:
LGTM other than the curlies

janekdb committed:
Group Console and AnsiColor entities and add usage examples

 - Group AnsiColor and Console members
 - Add Console IO redefinition examples
 - Add Console control code examples
 - Add color swatches to AnsiColor
 - Add AnsiColor control code examples
 - Link to StdIn for reading
 - Fix link syntax for java.util.Formatter
 - Fix withErr method example

janekdb commented:
@lrytz - Fully curly strings restored.
@soc - I can see that less curly braces would be preferable if you have become accustomed to this more streamlined format. I think for tutorial purposes the fully curlies have an advantage when scanning the code.

All changes done. Please re-review.

lrytz commented:
LGTM!
